src/main/java/com/mraof/minestuck/item/loot/MSLootEvents.java

```diff
@@ -2,28 +2,137 @@
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.level.storage.loot.LootPool;
 import net.minecraft.world.level.storage.loot.entries.LootTableReference;
 import net.minecraftforge.event.LootTableLoadEvent;
+import net.minecraftforge.event.village.VillagerTradesEvent;
+import net.minecraftforge.event.village.WandererTradesEvent;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 import net.minecraftforge.fml.common.Mod;
 
+import java.util.List;
+import java.util.Locale;
+import java.util.Random;
 import java.util.Set;
 
 @Mod.EventBusSubscriber(modid = Minestuck.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
 public class MSLootEvents
 {
-	private static final Set<ResourceLocation> LOOT_INJECT = Sets.newHashSet(BuiltInLootTables.SIMPLE_DUNGEON, BuiltInLootTables.ABANDONED_MINESHAFT, BuiltInLootTables.DESERT_PYRAMID, BuiltInLootTables.JUNGLE_TEMPLE, BuiltInLootTables.WOODLAND_MANSION, BuiltInLootTables.UNDERWATER_RUIN_BIG, BuiltInLootTables.SPAWN_BONUS_CHEST);
+	private static final Set<ResourceLocation> BLANK_DISK_DUNGEON_LOOT_INJECT = Sets.newHashSet(BuiltInLootTables.SIMPLE_DUNGEON, BuiltInLootTables.ABANDONED_MINESHAFT, BuiltInLootTables.DESERT_PYRAMID, BuiltInLootTables.JUNGLE_TEMPLE, BuiltInLootTables.WOODLAND_MANSION, BuiltInLootTables.UNDERWATER_RUIN_BIG, BuiltInLootTables.SPAWN_BONUS_CHEST);
+	private static final Set<ResourceLocation> SBURB_CODE_LIBRARY_LOOT_INJECT = Sets.newHashSet(BuiltInLootTables.STRONGHOLD_LIBRARY, BuiltInLootTables.LIBRARIAN_GIFT, BuiltInLootTables.STRONGHOLD_CORRIDOR);
 	
 	@SubscribeEvent
 	public static void onLootLoad(LootTableLoadEvent event) //created using Upgrade Aquatic "LootEvents" and Mystical World "LootHandler" for reference
 	{
-		if(LOOT_INJECT.contains(event.getName()))
+		inject(event, BLANK_DISK_DUNGEON_LOOT_INJECT, MSLootTables.BLANK_DISK_DUNGEON_LOOT_INJECT, "blank_disk_dungeon_inject");
+		inject(event, SBURB_CODE_LIBRARY_LOOT_INJECT, MSLootTables.SBURB_CODE_LIBRARY_LOOT_INJECT, "sburb_code_library_inject");
+	}
+	
+	private static void inject(LootTableLoadEvent event, Set<ResourceLocation> lootTableSet, ResourceLocation injectionLootTable, String address)
+	{
+		if(lootTableSet.contains(event.getName()))
 		{
-			LootPool pool = LootPool.lootPool().add(LootTableReference.lootTableReference(MSLootTables.DUNGEON_LOOT_INJECT)).name("dungeon_loot_inject").build();
+			LootPool pool = LootPool.lootPool().add(LootTableReference.lootTableReference(injectionLootTable)).name(address).build();
 			event.getTable().addPool(pool);
 		}
 	}
+	
+	@SubscribeEvent
+	public static void addCustomVillagerTrade(VillagerTradesEvent event)
+	{
+		Int2ObjectMap<List<VillagerTrades.ItemListing>> trades = event.getTrades();
+		
+		if(event.getType() == VillagerProfession.CARTOGRAPHER)
+		{
+			trades.get(1).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 8), new ItemStack(Items.COMPASS), createFrogTempleMap(villager), 12, 7, 0.2F));
+		}
```

Comment:
the world freezes up(lag) when a cartographer who has the frog temple map is first clicked, assumedly due to the trade loading. I use code I have assumed is pretty much identical to how vanilla handles the creation of the vanilla maps, but obviously they do not have any performance issues, its a one time thing per villager but it would be ideal to find a solution

Reply:
After pausing the game in debug mode when the world freezes up, it was [in the middle of finding the minimum block for the frog temple](https://github.com/Dweblenod/Minestuck/blob/7b470ae0b9b7a326e6a2a4e565879883e68e8168/src/main/java/com/mraof/minestuck/world/gen/structure/FrogTemplePiece.java#L46). There's a good reason to think that we need to rewrite the code for height displacement of that frog temple piece into something that performs better.

Reply:
is that something you think is appropriate to do in this PR?

Reply:
Not necessarily part of this PR as it doesn't already touch that part of the code, but it is a good idea to fix it before the next major release.

Reply:
Then I will resolve this particular conversation